### ISSUE TITLE
The running-from-SVN header is buggy

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#3526]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)| @thijstriemstra reported|
|-|-|
|Trac ID|trac#3526|
|Type|defect|
|Created|2008-11-09 17:41:17Z|
|Branch|https://github.com/twisted/twisted/tree/running-from-svn-3526-2|

The running-from-SVN header is buggy because it assumes a sub-project ```bin``` directory. Scripts for twisted core (twistd, trial) need one fewer reference to ```os.pardir```.

Attachments:

* [normalize.preamble.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/d30/d30c97fec8e99b50fc5e6b2d36197137cc93ee09/64dab19765e442f17b522784ecebc4cf85780525.patch) (16319 bytes) - added by techtonik on 2011-04-08 19:23:40Z - alt.patch

<details><summary>Searchable metadata</summary>

```
trac-id__3526 3526
type__defect defect
reporter__thijs thijs
priority__normal normal
milestone__ 
branch__branches_running_from_svn_3526_2 branches/running-from-svn-3526-2
branch_author__exarkun exarkun
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1226252477000000 1226252477000000
changetime__1309438636000000 1309438636000000
version__None None
owner__exarkun exarkun
cc__exarkun cc__thijs cc__jesstess cc__ivank
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_1"></a>|@exarkun commented|
|-|-|

I'm pretty sure the ticket description is inaccurate.  However, for subproject scripts, it needs one **more** `os.pardir`.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_2"></a>|@exarkun commented|
|-|-|

(In [25396]) Branching to 'scripts-3526'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_3"></a>|@exarkun removed owner|
|-|-|

I put an idea into the branch.  Some things to consider:

  1. Do we actually care about this functionality, or can we throw it away?
  1. Can we make this work properly with packaging?  How?
  1. How much testing is required?

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_4"></a>|@glyph set owner to @exarkun|
|-|-|

Replying to [exarkun](#note_3):
> I put an idea into the branch.  Some things to consider:

>   1. Do we actually care about this functionality, or can we throw it away?

The functionality that I care about here is a new user being able to get started with Twisted, having nothing but the tarball, and not installing it; not setting any environment variables or hacking `sys.path`.

The preamble, as such, may not be sufficient.  Nowadays, Twisted depends on zope.interface, and neither the checkout nor the downloadable tarball includes an importable version of zi, so the user must download and install that before anything would work.  I'd prefer that this work, but making this work right in the general case is basically making `easy_install` work correctly, then making Twisted work with it, and there's a lot of ground between here and there.

Unless we get rid of subproject subdirectories, there's no particular reason to keep this code there.

The other functionality `bin/` provides, in kind of a backwards way (since it predates combinator) is to allow combinator to recognize scripts that are part of Twisted and make it work when I type `twistd` or `trial` at the command line.  We could easily remove the preamble without breaking that.

>   2. Can we make this work properly with packaging?  How?

I'm not really sure what you mean by "properly".  Certainly you can't put a `_preamble.py` into `/usr/bin`.

My inclination would be to make `trunk/bin` be a convenience only for development from a checkout, and generate the actual scripts/ or bin/ from separate metadata when the releases are generated.  The shebang lines, for example, should be shebang lines for whatever interpreter is used to do the installation, not hard-coded to `/usr/bin/python`.  And on Windows there shouldn't even be a `bin/`, there should be a `scripts/`.

>   3. How much testing is required?

I think this depends on what the actual requirements for the purpose of these files is.  I have made some tools that work based on their presence, but I frankly have no idea how they relate to the release toolchain.  Someone else who has an idea of why they are as they are now, rather than the reason that the `bin/` directory was added in the first place, should probably comment on that before we decide.

Tests should verify that the scripts serve whatever purpose that they are supposed to serve; it wouldn't be terribly meaningful to test these without a tightly specified environment, so I think some `spawnProcess`-ing might be called for.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_5"></a>|@exarkun commented|
|-|-|

Replying to [glyph](#note_4):
> Replying to [exarkun](#note_3):
> > I put an idea into the branch.  Some things to consider:
>
> >   1. Do we actually care about this functionality, or can we throw it away?
>
> The functionality that I care about here is a new user being able to get started with Twisted, having nothing but the tarball, and not installing it; not setting any environment variables or hacking `sys.path`.
>
> The preamble, as such, may not be sufficient.  Nowadays, Twisted depends on zope.interface, and neither the checkout nor the downloadable tarball includes an importable version of zi, so the user must download and install that before anything would work.  I'd prefer that this work, but making this work right in the general case is basically making `easy_install` work correctly, then making Twisted work with it, and there's a lot of ground between here and there.
>
> Unless we get rid of subproject subdirectories, there's no particular reason to keep this code there.
>
> The other functionality `bin/` provides, in kind of a backwards way (since it predates combinator) is to allow combinator to recognize scripts that are part of Twisted and make it work when I type `twistd` or `trial` at the command line.  We could easily remove the preamble without breaking that.
>

It sounds like we care, but the functionality isn't actually provided now, as the _running-from-SVN_ header is not sufficient to provide it.  So we should get rid of the preamble everywhere.

One other thing, though.  The preamble includes code to add the working directory to sys.path - sometimes.  That sounds like a different feature than the one you described (and one I've disliked for some time, due to its inconsistent availability, normal user vs root, posix vs windows).

> >   2. Can we make this work properly with packaging?  How?
>
> I'm not really sure what you mean by "properly".  Certainly you can't put a `_preamble.py` into `/usr/bin`.
>
> My inclination would be to make `trunk/bin` be a convenience only for development from a checkout, and generate the actual scripts/ or bin/ from separate metadata when the releases are generated.  The shebang lines, for example, should be shebang lines for whatever interpreter is used to do the installation, not hard-coded to `/usr/bin/python`.  And on Windows there shouldn't even be a `bin/`, there should be a `scripts/`.

This basically sounds good to me, and it's really getting into the details of what I was asking.  Having `trunk/bin` be a developer-only convenience with some other mechanism to provide scripts for installations sounds great to me.  Can we overcome the technical obstacles to accomplish that?  I feel like the answer is no (at least with the current primitive packaging tools we're using).

I'm interested in finding a way to resolve this ticket without re-implementing distutils, setuptools, and aap from scratch.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_6"></a>|@exarkun commented|
|-|-|

(In [29724]) Branching to 'running-from-svn-3526'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_7"></a>|@exarkun removed owner|
|-|-|

I kinda forgot about the original branch here.  That was stupid.  But at least it only took about 5 minutes to re-implement everything in it.  I also added tests this time.

To sum up a few things:

> 1. Do we actually care about this functionality, or can we throw it away?

Yes, we care.  So now it's unit (well, sort of) tested

> 2. Can we make this work properly with packaging? How?

_We_ can't because we don't make the packages.  But `_preamble` is now imported such that if it isn't found, the script will just carry on.  This should be the correct behavior for the installed case.  Twisted is already in the path (that's the point of installation).

So, in particular:

> Certainly you can't put a _preamble.py into /usr/bin.

doesn't matter, because we don't want to do that, and now we don't need to.

Perhaps eventually we can generate the scripts at sdist (or something?) time and leave this out entirely.  That would be nice.

> 3. How much testing is required?

I tested enough to demonstrate that nearby vcs checkout of Twisted is used.  I didn't add any tests for the installed case, which means there's no tests to verify that the `ImportError` is being discarded.

It's _probably_ possible to come up with some tests that exercise the installed case.  I'll think about it a bit, in case anyone wants such tests before this is merged.

One thing that's presently _not_ in the branch at all anymore is the special non-root handling on POSIX.  I should probably put that back.  But... tests... yea.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_8"></a>|@exarkun commented|
|-|-|

(In [29843]) Update the running-from-SVN header documentation

refs [#3526](https://github.com/twisted/twisted/issues/3526)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_9"></a>|@exarkun commented|
|-|-|

(In [29844]) get rid of the extra list item

refs [#3526](https://github.com/twisted/twisted/issues/3526)

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_10"></a>|@thijstriemstra set owner to @exarkun|
|-|-|

 - in the description it mentions 'UNIX like', this should be 'UNIX-like'
 - maybe rewrite 'operating systems we use the' to 'operating systems, we use the', so adding a comma.
 - the word 'env' in should be wrapped in <code></code>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_12"></a>|@exarkun removed owner|
|-|-|

Those changes don't appear to be related to this ticket.

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_13"></a>|@thijstriemstra commented|
|-|-|

Replying to [exarkun](#note_12):
> Those changes don't appear to be related to this ticket.

Why not? That section describes the running-from-svn-header, I'm confused. If you don't want the docs to be improved than let me know.

**Comment by twisted-trac** [comment]
|<img alt="Screwtape's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Screwtape set owner to Screwtape|
|-|-|

**Comment by twisted-trac** [comment]
|<img alt="Screwtape's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_15"></a>|Screwtape set owner to @exarkun|
|-|-|

Replying to [thijs](#note_13):
> Why not? That section describes the running-from-svn-header, I'm confused. If you don't want the docs to be improved than let me know.

Looking closely, it seems that section describes module boilerplate in general; the paragraph that exarkun changed was the paragraph about running-from-SVN-header, the bit you're talking about was the previous paragraph about the hash-bang line.

As for my review, it generally looks good, except:

 1. Running tests on CentOS 5, I get crazy "error in processing external entity reference at line 239, column 2" exceptions from the Lore tests, but I rather suspect that's some misconfiguration on my end. Still, build results from py2.4 builders would be good to see.
 1. Running tests on Fedora 13, all the tests in `twisted.conch.test.test_scripts` fail because I don't have PyASN.1 installed. Running "bin/conch/conch --version" gives me a traceback rather than the "Twisted version: 10.1.0" that the tests are looking for. I guess you need to apply whatever `.skip` attribute is used to skip the rest of the Conch tests to these tests too.
 1. Needs a news file.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_16"></a>|@exarkun commented|
|-|-|

(In [30129]) Skip these tests if a dependency (PyCrypto or pyasn1 or Tkinter) is missing 

refs [#3526](https://github.com/twisted/twisted/issues/3526)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_17"></a>|@exarkun commented|
|-|-|

(In [30130]) Really skip tkconch if necessary

refs [#3526](https://github.com/twisted/twisted/issues/3526)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="jesstess's avatar" src="https://avatars.githubusercontent.com/u/188336?s=50" width="50" height="50">](https://github.com/jesstess)<a name="note_19"></a>|@jesstess set owner to @exarkun|
|-|-|

Thanks for sticking with this, exarkun.

 * [buildbot](http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/running-from-svn-3526) has an exciting mélange of issues.
 * bin/tapconvert didn't get a test.
 * Does _preamble.py need a copyright?

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|<img alt="techtonik's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_21"></a>|techtonik commented|
|-|-|

I need this patch to run scripts on Windows from checkout of /twisted/ for further development. [#5041](https://github.com/twisted/twisted/issues/5041) was marked as duplicate of this issue. Seems like this issue [#3526](https://github.com/twisted/twisted/issues/3526) got overcomplicated - tl;dr, but the patch fixes original user story mentioned in description.

Please, review.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_22"></a>|@exarkun commented|
|-|-|

(In [31575]) Branching to 'running-from-svn-3526-2'

**Comment by twisted-trac** [comment]
|<img alt="techtonik's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_23"></a>|techtonik commented|
|-|-|

Thanks for working on this.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_24"></a>|@exarkun set owner to @exarkun<br>@exarkun set status to `assigned`|
|-|-|

not really in review right now, branch has an issue with setuptools interaction, the cause of which isn't quite clear

Thanks for your patch, techtonik, but I'd rather continue working on the solution in the branch than abandon it and start from scratch in a different direction ignoring all of the issues that came up earlier during development of the branch.

**Comment by twisted-trac** [comment]
|<img alt="techtonik's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_25"></a>|techtonik commented|
|-|-|

What's wrong with setuptools?

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_26"></a>|@exarkun commented|
|-|-|

Two tests fail in this branch when run using the setuptools console entrypoint for trial, <http://buildbot.twistedmatrix.com/builders/debian-easy-py2.5-epoll/builds/1170>.  Something about how the trial script is launched causes a difference in the result of sys.exc_info() so that there is an exception returned from it when one is not expected.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_27"></a>|@exarkun removed owner<br>@exarkun set status to `new`|
|-|-|

This wasn't really a setuptools issue.  It was an installed case issue.  Too accustomed to finding fault with setuptools. ;)

I fixed this by adding `sys.exc_clear()` to the use-preamble boilerplate.  [Latest build results](http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/running-from-svn-3526-2).

I also added a test for the mailmail script change, along the lines of the other script tests already added.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_28"></a>|@therve set owner to @exarkun|
|-|-|

1. Some of the test files (twisted/lore/test/test_scripts.py, twisted/mail/test/test_scripts.py, twisted/conch/test/test_scripts.py) are missing blank lines between the imports and the test class.

Looks good, please merge.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_29"></a>|@exarkun commented|
|-|-|

(In [31709]) vertical whitespace

refs [#3526](https://github.com/twisted/twisted/issues/3526)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_30"></a>|@exarkun commented|
|-|-|

on IRC ivan pointed out that this removes the sys.path.insert(0, CWD) behavior from trial and twistd.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_31"></a>|@therve removed owner|
|-|-|

I did something.

**Comment by twisted-trac** [comment]
|<img alt="techtonik's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_32"></a>|techtonik commented|
|-|-|

This makes me wonder if `trial` and `twisted` (and maybe other scripts as well) should explicitly tell user where are they called from?

**Comment by twisted-trac** [comment]
|<img alt="ivank's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_33"></a>|ivank commented|
|-|-|

1. On IRC, I mentioned that this needs a NEWS entry:

twisted/topfiles/3526.bugfix: "trial and twistd now add the current directory to sys.path even when running as root or on Windows.  mktap, tapconvert, and pyhtmlizer no longer add the current directory to sys.path."

2. Please remove twisted/topfiles/3526.feature, because a refactoring of some code is not something that users care about. (It's not a new feature either.)

3. twisted/python/dist.py's `getScripts` lists .pyc in `specialExclusion`.  This neglects `.pyo` and Jython's `$py.class`.  Please add those, or make it exclude everything that starts with '_preamble' (note no trailing dot).  `test_excludedPreamble` needs to be updated too.

4. `twisted/conch/test/test_scripts.py` is still missing some whitespace after the imports.

Otherwise, looks good; merge if those are fixed.

**Comment by twisted-trac** [comment]
|<img alt="ivank's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_34"></a>|ivank commented|
|-|-|

techtonik: Do you mean you want every script like trial to announce "I'm in /usr/bin/trial"? The `which` command is handy for finding this information. Or do you mean you want them to announce `sys.path`? (This is pretty deterministic and probably doesn't need to be announced. Users with a lot of .egg directories from setuptools will see a lot of line noise.)

**Comment by twisted-trac** [comment]
|<img alt="techtonik's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_35"></a>|techtonik commented|
|-|-|

ivank: I want to know which trial is running when I execute the script. Like --version can show that trial tail is imported from \Python25\lib\something or from local checkout or from current directory.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_36"></a>|@exarkun commented|
|-|-|

> This neglects .pyo and Jython's $py.class.

This really only _needs_ to work with the documented release process.  We don't do releases with python -O or Jython.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_37"></a>|@glyph commented|
|-|-|

Replying to [exarkun](#note_36):
> > This neglects .pyo and Jython's $py.class.
> 
> This really only _needs_ to work with the documented release process.  We don't do releases with python -O or Jython.

True enough, is it really a good idea to have another ad-hoc list which repeats the static '.svn' string again?  It seems like we could just re-use `_filterNames` and get the same effect with less code.

This wouldn't fix the problem with `$py.class`, but it would at least give us just _one_ list to update when we want to fix it in the future.

**Comment by twisted-trac** [comment]
|<img alt="ivank's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_38"></a>|ivank commented|
|-|-|

I invite you to ignore my #3 entirely, if you wish (and maybe open a new ticket for refactorings).

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)|@therve set owner to @therve|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_40"></a>|@exarkun set owner to @exarkun|
|-|-|

There are some buildbot failures because the `sys.path` tests think that a Python process can reliably produce no output, but that's totally false. :)

eg, on the OS X builder, the child process produces:

```
/System/Library/Frameworks/Python.framework/Versions/2.6/Extras/lib/python/zope/__init__.py:1: 
    UserWarning: Module twisted was already imported from /Users/exarkun/Run/twisted-slave/osx10.6-py2.6-select/Twisted/twisted/__init__.pyc,
        but /System/Library/Frameworks/Python.framework/Versions/2.6/Extras/lib/python is being added to sys.path
  __import__('pkg_resources').declare_namespace(__name__)
```

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_41"></a>|@exarkun commented|
|-|-|

(In [31913]) Check stdout for some known-good output instead of verifying that there is no stderr output

refs [#3526](https://github.com/twisted/twisted/issues/3526)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_42"></a>|@exarkun commented|
|-|-|

I twiddled the test a little bit.  [Build results](http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/running-from-svn-3526-2).

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_43"></a>|@glyph commented|
|-|-|

Replying to [exarkun](#note_40):
> There are some buildbot failures because the `sys.path` tests think that a Python process can reliably produce no output, but that's totally false. :)
> 
> eg, on the OS X builder, the child process produces:
> 
> ... Module twisted was already imported from ... but ... is being added to sys.path ...

In case some poor soul comes across this ticket while searching for this problem (and as a potential solution for similar tests) this can be fixed, on multiple platforms, by `python setup.py egg_info` in the Twisted directory under test.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_45"></a>|@glyph set owner to @glyph<br>@glyph set status to `assigned`|
|-|-|

Reviewing.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_46"></a>|@glyph set owner to @exarkun<br>@glyph set status to `new`|
|-|-|

Although the buildbots pass the tests for this branch, on my machine, I see this traceback (only when running the full suite):
```
Traceback (most recent call last):
  File "twisted/trial/runner.py", line 570, in loadPackage
    module = modinfo.load()
  File "twisted/python/modules.py", line 383, in load
    return self.pathEntry.pythonPath.moduleLoader(self.name)
  File "twisted/python/reflect.py", line 464, in namedAny
    topLevelPackage = _importAndCheckStack(trialname)
  File "twisted/python/reflect.py", line 400, in _importAndCheckStack
    return __import__(importName)
  File "twisted/mail/test/test_imap.py", line 3375, in <module>
    class NewFetchTestCase(unittest.TestCase, IMAP4HelperMixin):
  File "twisted/mail/test/test_imap.py", line 3486, in NewFetchTestCase
    locale.setlocale(locale.LC_ALL, currentLocale)
  File ".../lib/python2.6/locale.py", line 494, in setlocale
    return _setlocale(category, locale)
locale.Error: unsupported locale setting

twisted.mail.test.test_imap
```
This occurs because, for some reason I can't divine, importing `Tkinter`, on a mac, _only when you are running under a GUI_, causes the locale to be set to an invalid value ("`en_US.UTF8`", to be precise, as opposed to the correct "`en_US.UTF-8`").  Actually many of our uses of the locale module look to be at odds with the observed behavior on a couple of systems I've looked at.  Most importantly, the value returned from `getlocale` can't necessarily safely be passed to `setlocale`, even if it's actually a supported locale.  So this needs to be addressed somehow, albeit perhaps a separate ticket would be good for attacking this, since it's mostly an accident that this doesn't occur in trunk.

More relevant to the actual content of the branch: it would be nice to get rid of the subdirectories in bin/.  For one thing, one reason for having a '`bin`' directory is to have an entry to put on `$PATH` during development, and subdirectories break it.  For another, we now have a preamble-preamble, which is a whole separate maintenance headache.  Of course, then we'd need an addition to our release management tools to declare the separation of scripts into different projects.  I think the thing to do here is to file a new ticket.

Please address these two issues to your satisfaction and land, unless the solution to the first one is complex enough to warrant a second review.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_47"></a>|@exarkun commented|
|-|-|

The locale issue is fixed in r32199

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_48"></a>|@exarkun commented|
|-|-|

Filed [#5168](https://github.com/twisted/twisted/issues/5168) for the bin/ subdirectories.  I want to avoid touching it for now because it means teaching the release tools a new trick.  It would definitely be a nice simplification though.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_49"></a>|@exarkun commented|
|-|-|

(In [32201]) assertEquals -> assertEqual

refs [#3526](https://github.com/twisted/twisted/issues/3526)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_50"></a>|@exarkun set status to `closed`|
|-|-|

(In [32202]) Merge running-from-svn-3526-2 

Author: exarkun
Reviewer: glyph, Screwtape, jesstess, therve, ivank
Fixes: [#3526](https://github.com/twisted/twisted/issues/3526)

Factor the duplicate `sys.path` manipulation done by the command line scripts
to share a single implementation.


Also change trial and twistd to always add the current working directory to
`sys.path` (previously it would be added unless running as root or on Windows).
